### PR TITLE
Fixed the Hungarian tax id regex

### PR DESCRIPTION
**Pull request by abdallahbeshi**
## Description

As per the technical documentation, the HU tax id must only consist of 8 numbers
image

So we extended the tax id validation to validate based on the client usage, either by allowing the country code prefix in the tax id or without.

## Type of change
- [x] Bug fix.
- [ ] Feature.
- [ ] Consolidation.

## Checklist
- [ ] Is breaking change.
- [ ] Documentation updated.
- [ ] Tests included (please specify cases).

### CODE REVIEW

**Comment by danielchorous** [comment]
@abdallahbeshi we should also fix BE crashing in case of an unsuccessful request. Let's have a call about this?

**Comment by abdallahbeshi** [comment]
> @abdallahbeshi we should also fix BE crashing in case of an unsuccessful request. Let's have a call about this?

We actually want it to throw an error if the library is not handling some response so we get notified and fix it.

**Comment by KaliCZ** [comment]
@abdallahbeshi Correct. But now we know that there is a possibility for the server to return this type of response. So we should handle this type of response.

**Comment by abdallahbeshi** [comment]
> 

Yea we knew that before, the server returned an invalid request because we sent an invalid XML.

I can make it handle the "invalid request" response we get from the server, but i don't think that's a priority now, so pls create a YT

**Comment by KaliCZ** [comment]
What do you mean invalid XML? the XML schema didn't change right? This value was valid before and now the value is invalid. But the web service is still the same with the same XML schema.

**Comment by KaliCZ** [comment]
Or like it was crashing the whole time.

**Comment by abdallahbeshi** [comment]
> What do you mean invalid XML? the XML schema didn't change right? This value was valid before and now the value is invalid. But the web service is still the same with the same XML schema.

The XML schema didn't change, it's just that we got lucky that no one ever provided an invalid HU tax ID :D

File: src/Core/Mews.Fiscalizations.Core/Model/Countries/Countries.cs [hunk 1]
```diff
@@ -110,7 +110,7 @@ static Countries()
                 HolySee = NonEuropean("VA"),
                 Honduras = NonEuropean("HN"),
                 HongKong = NonEuropean("HK"),
-                Hungary = European("HU", "^(HU)?[0-9]{8}$"),
+                Hungary = European("HU", "^[0-9]{8}$"),
```
Comment (KaliCZ) [inline src/Core/Mews.Fiscalizations.Core/Model/Countries/Countries.cs:113]: The HU has to be allowed. So we can create a second regex and make Hungary use the second one. But generally, it's OK to provide the country code in TaxID.

File: src/Hungary/Mews.Fiscalizations.Hungary/Models/Invoice/Receiver.cs [hunk 2]
```diff
@@ -15,7 +15,7 @@ public Receiver(Company company)
         {
         }
 
-        public static ITry<Receiver, Error> LocalCompany(TaxpayerIdentificationNumber taxpayerId, Name name, SimpleAddress address)
+        public static ITry<Receiver, Error> LocalCompany(string taxpayerId, Name name, SimpleAddress address)
```
Comment (KaliCZ) [inline src/Hungary/Mews.Fiscalizations.Hungary/Models/Invoice/Receiver.cs:18]: Does this simplify the usage?

File: src/Hungary/Mews.Fiscalizations.Hungary/Models/Invoice/LocalCompany.cs [hunk 3]
```diff
@@ -19,12 +18,9 @@ private LocalCompany(TaxpayerIdentificationNumber taxpayerId, Name name, SimpleA
 
         public SimpleAddress Address { get; }
 
-        public static ITry<LocalCompany, Error> Create(TaxpayerIdentificationNumber taxpayerId, Name name, SimpleAddress address)
+        public static ITry<LocalCompany, Error> Create(string taxpayerId, Name name, SimpleAddress address)
```
Comment (KaliCZ) [inline src/Hungary/Mews.Fiscalizations.Hungary/Models/Invoice/LocalCompany.cs:21]: Same question.

File: src/Core/Mews.Fiscalizations.Core/Model/Countries/EuropeanUnionCountry.cs [hunk 4]
```diff
@@ -5,15 +5,18 @@ namespace Mews.Fiscalizations.Core.Model
 {
     public class EuropeanUnionCountry
     {
-        internal EuropeanUnionCountry(string alpha2Code, Regex taxpayerNumberPattern)
+        internal EuropeanUnionCountry(string alpha2Code, Regex regexWithCountryCodePrefix, Regex regexWithoutCountryCodePrefix)
         {
             Alpha2Code = alpha2Code;
-            TaxpayerNumberPattern = taxpayerNumberPattern;
+            RegexWithCountryCodePrefix = regexWithCountryCodePrefix;
+            RegexWithoutCountryCodePrefix = regexWithoutCountryCodePrefix;
```
Comment (KaliCZ) [inline src/Core/Mews.Fiscalizations.Core/Model/Countries/EuropeanUnionCountry.cs:12]: Right, but we need to name it better. This class is a country. It's not obvious, what the regex is for.
So maybe `taxpayerNumberPattern` and `taxpayerNumberPatternWithoutCountryCodePrefix`?

**Review by KaliCZ — APPROVED** [review]

File: src/Hungary/Mews.Fiscalizations.Hungary/Models/Invoice/LocalCompany.cs [hunk 5]
```diff
@@ -19,12 +18,9 @@ private LocalCompany(TaxpayerIdentificationNumber taxpayerId, Name name, SimpleA
...
         {
-            var localTaxPayerId = taxpayerId.ToOption().Where(i => i.Country.Equals(Countries.Hungary));
-            return localTaxPayerId.ToTry(_ => new Error($"{nameof(TaxpayerIdentificationNumber)} must be a Hungarian taxpayer number.")).Map(
-                i => new LocalCompany(i, name, address)
-            );
+            return TaxpayerIdentificationNumber.Create(Countries.Hungary, taxpayerId).Map(n => new LocalCompany(n, name, address));
```
Comment (KaliCZ) [inline src/Hungary/Mews.Fiscalizations.Hungary/Models/Invoice/LocalCompany.cs:23]: Shouldn't this actually say AllowsPrefix: false?

File: src/Core/Mews.Fiscalizations.Core/Model/Legal/EuropeanUnionTaxpayerIdentificationNumber.cs [hunk 6]
```diff
@@ -14,11 +14,20 @@ private EuropeanUnionTaxpayerIdentificationNumber(EuropeanUnionCountry country,
...
+            EuropeanUnionCountry country,
+            string taxpayerNumber,
+            bool isCountryCodePrefixAllowed = true)
         {
             return ObjectValidations.NotNull(country).FlatMap(c =>
             {
-                var validNumber = StringValidations.RegexMatch(taxpayerNumber, country.TaxpayerNumberPattern);
+                var validNumber = StringValidations.RegexMatch(
+                    value: taxpayerNumber,
+                    pattern: isCountryCodePrefixAllowed.Match(
```
Comment (KaliCZ) [inline src/Core/Mews.Fiscalizations.Core/Model/Legal/EuropeanUnionTaxpayerIdentificationNumber.cs:26]: I'd just create a variable pattern and leave this call as a single-line

File: src/Hungary/Mews.Fiscalizations.Hungary/Models/Invoice/Receiver.cs (hunk 2 above)
Comment (abdallahbeshi) [reply src/Hungary/Mews.Fiscalizations.Hungary/Models/Invoice/Receiver.cs:18]: Yea it should as you don't have to validate the tax id and then the local company
there is an example in the modified tests

**Review by KaliCZ — APPROVED** [review]

**Pull request merged**